### DISCUSSION TITLE
How smart switch shall work when bound?

### DISCUSSION THREAD

**Discussion by grafalex82** [Koenkk/zigbee2mqtt#20743] [General]
Hi All,

Imagine a typical smart switch, like Aqara or Tuya, consider a 1-gang switch for simplicity. It has a button and a relay. Despite the 'Switch' word in the name, from Zigbee standpoint it is a Light (or a relay) - a server type device, that can handle On/Off/Toggle commands, set its relay state accordingly, and report current state to the coordinator if changed. The button here is just a way to toggle light state.

At the same time you may have another Zigbee light, that you want to control with your switch. In this case switch is a controller (client) type device. It sends On/Off/Commands to the light. Logic relationships between the light and the switch can be set using Home Assistant (or a similar tool),. Or you may employ Zigbee binding between the switch and the light, so that this connection works even without the coordinator. 

**Question:** What is the moment when the switch (which was actually a Light in the first example) becomes a controller type device? How the switch understand which mode it shall work in?

Here is the context on my question. I am developing a smart switch. The selection between the Light and Controller mode is done by the fact if any other Light is bound: if the switch is bound to a light - the switch works like a client switch, and sends commands to the Light; if nothing is bound - it works like a relay, and accept commands from other switches. The problem starts when my device is added to the group - zigbee2mqtt sets up reporting to the coordinator, and binds device's endpoints. This totally breaks logic of my device, and I am trying to figure out what is the right behavior to implement.

**Comment by Koenkk** [comment]
>  if a light is bound to the device

I guess you mean: "if the switch is bound to a light" here?

> This totally breaks logic of my device, and I am trying to figure out what is the right behavior to implement.

I don't understand the problem fully, why does it break?

**Reply by grafalex82** [reply]
> I guess you mean: "if the switch is bound to a light" here?

Yes, correct. I also changed the original post to minimize a confusion.

> I don't understand the problem fully, why does it break?

Perhaps I need to make a wider explanation here. First let's talk about **Client type devices**. Let's start reasoning with this question: Suppose we have a controller (say, a button that generates commands to a light). Shall this button maintain its internal state (On/Off)?

If yes, it can generate On and Off commands to the light. At the same time there are at least 2 source of truth in this system regarding to the light state - the light, and the button. There is a risk that these state will get out of sync one day. For example the light is switched off, while the button still thinks it is on. In this case next command will be interpreted incorrectly (button will send Off command to the light that is off already). The problem may get even worse if you have several buttons that control a single light. 

This reasoning leads to a conclusion that the button shall NOT have an internal state (This reasoning matches the ZLL specs btw, where controller devices are client devices). Having no button internal state leads us to using the Toggle command to the light. In this case every button press will be treated by the light in the same way - the light state will be toggled. The light will never come out of sync with the button state because there is simply no button state. 


Now let's t talk about **Server devices**, or simply **Lights**. Things are relatively easy here - the device maintains its internal state, and accepts On/Off/Toggle commands. It may also report its state to subscribers. The device can report if a change happened, or a client may explicitly subscribe for an attribute (e.g. using Configure Reporting command). 

Things get more complicated with **Client+Server smart switches**:
- In some configurations it works like a Light - it handles the internal relay, maintains and reports its internal state, and accept commands from other switches.
- In other configurations it works like a controller button - it has no internal state, and sends commands Toggle to the bound Light

In my particular implementation the selection between the 2 modes is implemented by the fact the device is bound somewhere. If no binding - it is a light, if there are bindings - it is a button controlling some other light. Perhaps this logic is not correct, and I am actually trying to clarify if there are any other options.

Back to the question 'what exactly is broken?'. The problem begins when the switch is added to a group. Obviously in a group the switch can behave only as a server (Light). It has its state, which is reported to the coordinator. Perhaps not all lights do this, that is why Z2M explicitly configure reporting.

But at the same time Z2M also sets up a **binding** from the device to the coordinator for some reason. I disagree with this behavior for a number of reasons:
- binding **from** the server device looks meaningless, at least for commands. The server device does not generate any commands. Perhaps binding can be used to route reports to the coordinator, but why configure reporting could not be used instead?
- In my particular case the device switches to the Client mode. It generates Toggle commands due to reasoning above. This command works properly with other Lights, but Z2M does not hangle Toggle commands as state reports. 

While I disagree with Z2M implementation. I think it shall not perform binding when adding the device to a group. But perhaps I am not in a position to fight with it, as there could be other reasons to do this.

This all leads to a very initial question: how a smart switch shall behave when it is bound to another device or coordinator?

**Comment by Koenkk** [comment]
> Suppose we have a controller (say, a button that generates commands to a light). Shall this button maintain its internal state (On/Off)?

It shouldn't, it should send a `toggle` command, not `on` or `off`

> But at the same time Z2M also sets up a binding from the device to the coordinator for some reason. I disagree with this behavior for a number of reasons:

The reason for this is to get state updates when the device state is changed via the group. But if you don't want to have this behaviour, you can easily unbind this via the z2m frontend -> device -> binding tab.

**Reply by grafalex82** [reply]
> The reason for this is to get state updates when the device state is changed via the group.

Why regular tribute reports not enough? Z2M performs configure reporting for state attribute.
What kind of data does it expect through the bound transfer, reports or commands?

**Reply by Koenkk** [reply]
I'm wondering if the reporting also works without the binding?

**Reply by grafalex82** [reply]
Yes, reporting and binding are absolutely separate concepts. They even work on different OSI model levels.
- Configure Reporting is the ZCL feature is the mechanism to set up periodic reporting for an attribute. It can be also used to limit timings if the attribute value changes too often. 
- Binding is the Zigbee network level mechanism to deliver a message to a group of devices. Just broadcast a message to any device that is a part of the group. If can work even without ZCL involved, and any attibute defined.

The paragraph 2.5.7 of ZCL specification (Configure Reporting Command) does not obligate to use binding. The client device or coordinator simply subscribe for reports and get these reports directly as unicast message.

I do not have a big variety of devices laying around. I tested only on these:
- IKEA lamps - report On/Off attribute normally (I managed to set reporting each 30s)
- Xiaomi switch - report On/Off attributes normally, except it does not allow to set reporting frequently than 60s (default NXP SDK setting)
- Tuya temperature sensor rejected Configure Reporting commands as unsupported. But this is not a problem, as this sensor's aim is already to report temperature from time to time.

**Reply by Koenkk** [reply]
> IKEA lamps - report On/Off attribute normally (I managed to set reporting each 30s)

Does the reporting keep working when you remove the `onOff` bind to the coordinator?

**Reply by grafalex82** [reply]
I need to perform a few tests first. Is there a way to disable bind requests when setting up reporting?

**Reply by Koenkk** [reply]
No, but you can unbind it easily via the z2m frontend.

**Reply by grafalex82** [reply]
I ended up running z2m from sources, and tried commenting the bind() call in the bridge.ts, deviceConfigureReporting() function. 

Well, although in my picture of the world binding and reporting are separate things, I confirm that at least some devices (such as IKEA lamps) require building for reporting function. If I comment the bind() call and try to set up reporting, the lamp returns success, but reports are not emited.

Shall this approach (binding+reporting) be applied for all devices? Still not sure. Xiaomi Aqara can report even without binding. This leads to a question: can the set up reporting behavior be pushed on a converter level? Thus the decision whether to do binding or not could be programmed per device, rather than use binding for everything. What do you think?


And the original question is actual: How the smart switch shall behave when added to the group (and therefore being bound)?

**Reply by Koenkk** [reply]
> Thus the decision whether to do binding or not could be programmed per device, rather than use binding for everything. What do you think?

To do that properly you will need to kinda go to all devices (which is a huge amount so not feasible)

I'm wondering, since your use case is quite specific, can't you just unbind it after adding it to the group? I try to be a bit conservative with such fixes for such a specific use case.

**Reply by grafalex82** [reply]
> I'm wondering, since your use case is quite specific, can't you just unbind it after adding it to the group?

Can I do this automatically in the converter somehow?

Currently I am doing this manually. But as soon as my ultimage goal is an alternative firmware for Xiaomi Aqara switches, I do not want to bother users with additional actions they have to keep in mind when adding the switch to the group.

**Reply by Koenkk** [reply]
Yes, we have to add the `groupMembersChanged` to https://github.com/Koenkk/zigbee2mqtt/blob/master/lib/extension/onEvent.ts, then you can use it in the external converter `onEvent`.

**Reply by grafalex82** [reply]
Thank you for the info. That could be very helpful.

However I have a few updates on reporting vs binding topic. I am negatively surprized, but reporting on my device also does not work without binding (I appologize that I stated the opposite above). Moreover looks like this is done intentionally - the ZCL SDK code by NXP sends report using the bound transfer mode explicitly. So it looks like binding is really necessary. I have not found a special note in Zigbee specification about this, but the most relevant information is in section 2.2.2 of ZCL specification:
![image](https://github.com/Koenkk/zigbee2mqtt/assets/23584882/ee79fd16-7289-42a6-a6ee-6a82f6f95028)

I ended up introducing a user setting for my device that explicitly selects server or client mode of the switch (similar to event/command setting on some of the mass produced devices).
